Refactor/dynamic form structure

the-vampiire opened the pull request:
New Dynamic Form structure. Makes it easier to use just the pieces you need. Provides hooks for all related events in the DF process.

# Changes
- caught up with dev
- refactored VoyageApplication, Register, and WeeklyCheckin to use new Dynamic Form structure
- tested and working

### BREAKING
- now exports an object from `/DynamicForm`
- contains components: DynamicForm, DynamicFormWrapper, DynamicFormContainer
- contains utilities: dynamicFormMaker (func), questionComponents (obj)

### Usage
- import the component(s) you need from `/DynamicForm`

# Event hooks
## `onSubmit`
- accepted as prop in
  - `<DynamicForm ... />`
  - `<DynamicFormWrapper ... />`
- called when `Submit` button of DynamicForm is clicked
- receives `form_data` argument
### expected behavior
- perform mutation
- handle response data
- handle response error
### implementation
```js
handleSubmit = (form_data) => {
    const {
      client,
      mutation,
      purpose,
      version,
    } = this.props;
  
    const variables = {
      purpose,
      version,
      form_data,
    };

    client.mutate({ mutation, variables })
      .then(this.handleResponse)
      .catch(this.handleError);
  }
```

## `onResponse`
- accepted as prop in
  - `<DynamicForm ... />`
  - `<DynamicFormWrapper ... />`
- called when data is received back from the submit mutation
  - `onSubmit()` --> `onResponse(data)`
- receives `data` as an argument
### expected behavior
- process response data
- return a component (or `null`) to render 
### implementation
```js
if (response) return (
      onResponse
        ? onResponse(response)
        : <Redirect to="/" />
    );
```

## `onError`
- accepted as prop in
  - `<DynamicForm ... />`
  - `<DynamicFormWrapper ... />`
- called when an error is received back from the submit mutation
  - `onSubmit()` --> `onError(error)`
- receives `error` (error message) as an argument
### expected behavior
- process response error
- return a component (or `null`) to render 
### implementation
```js
if (error) return (
      onError
        ? onError(error)
        : <Error error={error} />
    );
```

# Components

## DynamicForm
### use case: all inclusive dynamic form
- queries for dynamic form given `purpose` and (optionally) `version`
- `hidden` input values can receive data via:
  - `hiddenData` object of `field`: `value` pairs 
  - `queryString` data in `field=value&field2=value2` pairs
- injects form data into DynamicFormWrapper
- optional custom mutation or uses default dynamicFormSubmit mutation
### props
```js
/**
 * @prop {string} purpose Dynamic Form purpose
 * --- OPTIONAL ---
 * @prop {number} version Dynamic Form version for given purpose
 * @prop {object} hiddenData object of values for hidden inputs
 * @prop {string} queryString query string data for hidden inputs
 * @prop {object} mutation alternative mutation (default submitDynamicForm)
 * @prop {func} onSubmit custom handler for submit
 * @prop {func} onResponse custom handler for mutation response data
 * @prop {func} onError custom handler for mutation error
 */
```

## DynamicFormWrapper
### use case: customize how the DF receives form and hidden data
- manages submission, response, and errors
- can be extended to provide custom behavior
### props
```js
/**
 * @prop {object} client Apollo Client instance for mutating
 * @prop {string} purpose form purpose (database collection)
 * @prop {number} version form version
 * @prop {array} questions form questions
 * 
 * -- OPTIONAL -- 
 * @prop {object} hiddenData optional object of hidden input field value(s)
 * @prop {func} onSubmit optional handler for submitting mutation
 * @prop {func} onResponse optional handler for response data
 * @prop {func} onError optional handler for error data
 */
```

## DynamicFormContainer
### use case: provide `questions` manually (simple dynamic form rendering)
- manages form_data state
- rendersr Question components (inputs)
### props
```js
/**
 * @prop {array} questions array of Question data objects for rendering
 * @prop {string} purpose Dynamic Form purpose (for LS form data persistence)
 * @prop {function} onSubmit wrapper callback for handling submit behavior
 */
```
### expected `questions` shape
```js
const questionShape = {
  id: PropTypes.string,
  text: PropTypes.string,
  subtext: PropTypes.string,
  input_type: PropTypes.string,
  field_name: PropTypes.string,
  options: PropTypes.arrayOf(
    PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  ),
  minlength: PropTypes.number,
  maxlength: PropTypes.number,
  data_type: PropTypes.string,
  tags: PropTypes.arrayOf(PropTypes.string),
};
```

thinktwice13 commented:
This looks awesome. Easy to follow.
My only note/question is that `dropdown_multiple` has its own input_type, but there is now only one `dropdown` questionComponent export whose `multiple` prop is hardcoded to false.
Similar thing with 1radio_special_badge`.
Both needed in voyage application

the-vampiire commented:
dropdown_multiple is not working. i replaced it with checkbox_2_column which is working. 

added voyage tier selection buttons

<img width="474" alt="screen shot 2018-08-28 at 2 54 54 pm" src="https://user-images.githubusercontent.com/25523682/44744173-7ce47080-aad2-11e8-91b5-0d4484e5b5a6.png">

the-vampiire commented:
TODO: we will have to redo the Form questions to use the new format. this had to be done anyways when we deploy but just a note that the form will not render without the new format

i tested it locally with the new format and it all works fine
